Functions for adding arbitrary unitary matrices to bosonic qiskit circuits, and a rework of cv_fockcounts

liu-zixiong opened the pull request:
Hi Tim,

I'm not sure what the etiquette is for pull requests, but I figured it might be good to list the changes I've been working on. Here are a couple of new functions that Kevin told me would be good to add.

1. Added _util.fockmap()_ with unit test. Allows user to generate matrices by specifying arbitrary mappings between a set of input and output fock states. 
2. Added _operators.gate_from_matrix_. Used by _cv_gate_from_matrix_ when appending _ParameterizedUnitaryGate_
3. Added _circuit.cv_gate_from_matrix_ with unit test. Converts user specified unitary matrices to gates in the circuit. Meant to be used with _fockmap()_ to generate any arbitrary gate. 
4. Added _CVCircuit_ property _qumode_qubits_indices_grouped_. Same as _qumode_qubits_indices_, but it groups the qubits that belong to same qumodes together. Returns nested list.
5. Created _util.newcounts()_ with unit test. Meant as a replacement to _util.cv_fockcounts_. Instead of taking in (counts, qubit_qumode_list), _newcounts_ only needs needs (circuit, result) as its arguments and does the rest. _newcounts_ depends on the __final_qumode_mapping_ auxiliary function, and __final_measurement_mapping_ that I lifted from the _Mthree_ package.
6. Integrated _newcounts()_ into _simulate()_ as an argument named _return_fockcounts_

tjstavenger-pnnl commented:
I'm definitely supportive of PRs and having someone review the changes before pushing to main. I don't want to be the only gatekeeper though -- if you've been working with Kevin on this work, you could have him review too (or instead of me).

I also appreciate the test cases -- definitely helpful while you're developing and also into the future to help ensure new changes don't break your code.

Is the copied Mthree code the same as this PyPI package? https://pypi.org/project/mthree/ Should we just add it as a dependency to the requirements.txt instead of copying in the code?

liu-zixiong commented:
Hi Tim, Kevin has been the person who suggested that we add these features to the repo but he hasn't actually looked at the actual way I've implemented his suggestions. But he has said that you're the primary developer for the repo, so I figured it might be good for you to have a look through as you'd have a higher chance of bumping into the pieces I've changed (especially with the new arg I added on simulate). For now I'll just add Kevin to the list of reviewers just in case he wants to take a look.

I'm currently working on a tutorial for error correction in bosonic qiskit using the binomial kitten code, and that tutorial will have examples of how one would use the features I've added. It's almost done, I just need to integrate your newest discretize code into the tutorial. But if you want a sense of how the features I've added would be used I can push the current version onto the repo so that you can take a look.

The copied Mthree code is exactly from that package. Kevin and I didn't think that it would be necessary to add the whole package in since I'm only using a single function, but if you disagree I can also replace the code with an import and add Mthree it as a dependency. Either way would work.

kevincsmith commented on c2qa/util.py:
One minor suggestion is to replace the name of "newcounts" with something that more clearly indicates its function. Something like "convert_counts" or "counts_to_fockcounts" may be clearer.

kevincsmith replied:
I might also suggest adding a warning under cv_fockcounts which points the user in the direction of the new, preferred method for converting the counts to a fock representation

liu-zixiong replied:
Thanks for the comments Kevin! Here are my adjustments:

>renamed "newcounts" to "counts_to_fockcounts"
>adjusted "return_fockcounts = True" in util.simulate()
>added a warning message under cv_fockcounts that says "The function ``c2qa.util.cv_fockcounts()`` is deprecated as of bosonic-qiskit v8.2. The function is replaced by counts_to_fockcounts(), which can be called using util.simulate()."

liu-zixiong replied:
Hmm, it seems that after setting "return_fockcounts = True" all of the unit tests are failing since simulate(return_fockcounts = True) will return a tuple of 3 values rather than 2. 

Would it be better to adjust all the unit tests? Or would it be better to set return_fockcounts=False by default?

liu-zixiong replied:
Thanks for the suggestions Tim! Here are the adjustments:
> `simulate()` returns tuple of 3 now

> Added `else` block and removed `continue`
To be clear, does this mean that you agree with adjusting all of the test functions so that they now unpack a tuple of 3? I'll get started on those changes once I have your confirmation.

liu-zixiong replied:
Thanks again Tim. Here are the changes. Feel free to merge the pull request anytime if there are no further comments.
> Within tests folder, replaced all instances of` state, result = c2qa.util.simulate()` with 
`state, result, _ = c2qa.util.simulate()`

> adjusted wigner.py functions to properly unpack when using simulate()

> adjusted`util.measure_all_xyz()` output to return
`(state_x, result_x, fockcounts_x), (state_y, result_y, fockcounts_y), (state_z, result_z, fockcounts_z)`

> adjusted `util.cv_ancilla_fock_measure()` line 51 from `_, result = simulate(circuit, shots=1)` to
` _, result, _ = simulate(circuit, shots=1)`

There is an opportunity to refactor util.cv_ancilla_fock_measure() to use the new fockcounts info, but as I'm unsure of what's going on in the function I've left it as is.

kevincsmith replied:
Thanks for making these changes @liu-zixiong. To add a quick comment on `util.cv_fock_measure()`, this is a method intended to carry out a hardware-realistic photon-number readout written by @EllaCrane. It uses an ancilla qubit to readout the photon number in binary bit-by-bit by using phase-kickback. So this is a method which is a bit separate from the idea of fockcounts, and can be left as-is (unless she sees a use for it to be incorporated).

kevincsmith reviewed (approved):
Thanks for all of the nice changes Zixiong. Everything looks good to me aside from the suggested renaming of `newcounts` and some sort of user feedback about `cv_fockcounts` no longer being the preferred method.

We should also consider making `return_fockcounts = True` the default option if we are sure it is working as intended.
